crates/matrix-sdk/src/widget/machine/mod.rs

```diff
@@ -25,23 +38,59 @@ pub(crate) use self::{
 #[cfg(doc)]
 use super::WidgetDriver;
 
-mod actions;
-mod events;
-mod openid;
-mod outgoing;
-
 /// No I/O state machine.
 ///
 /// Handles interactions with the widget as well as the `MatrixDriver`.
-pub(crate) struct WidgetMachine;
+pub(crate) struct WidgetMachine {
+    widget_id: String,
+    actions_sender: UnboundedSender<Action>,
+}
 
 impl WidgetMachine {
     /// Creates a new instance of a client widget API state machine.
     /// Returns the client api handler as well as the channel to receive
     /// actions (commands) from the client.
-    pub(crate) fn new() -> (Self, UnboundedReceiver<Action>) {
-        let (_tx, rx) = unbounded_channel();
-        (Self, rx)
+    pub(crate) fn new(
+        widget_id: String,
+        init_on_content_load: bool,
+    ) -> (Self, UnboundedReceiver<Action>) {
+        let (actions_sender, actions_receiver) = unbounded_channel();
+        let this = Self { widget_id, actions_sender };
+
+        if !init_on_content_load {
+            this.send_to_widget(RequestPermissions {});
+        }
+
+        (this, actions_receiver)
+    }
+
+    fn send_to_widget<T: ToWidgetRequest>(&self, to_widget_request: T) {
+        #[derive(Serialize)]
+        #[serde(tag = "api", rename = "toWidget", rename_all = "camelCase")]
+        struct ToWidgetRequestSerHelper<'a, T> {
+            widget_id: &'a str,
+            request_id: Uuid,
+            action: &'static str,
+            data: T,
+        }
+
+        let full_request = ToWidgetRequestSerHelper {
+            widget_id: &self.widget_id,
+            request_id: Uuid::new_v4(),
+            action: T::ACTION,
+            data: to_widget_request,
+        };
+        let serialized = match serde_json::to_string(&full_request) {
+            Ok(msg) => msg,
+            Err(e) => {
+                error!("Failed to serialize outgoing message: {e}");
```

Comment:
Do we have a span somewhere that will give a hint about context here? Otherwise, can you make the error message a little bit more specific please?

Reply:
Well, it's incredibly unlikely serialization would ever fail but sure, I can add the action as context.